session/session.go

```diff
@@ -1256,13 +1265,37 @@ func (s *session) SetProcessInfo(sql string, t time.Time, command byte, maxExecu
 	s.processInfo.Store(&pi)
 }
 
-func (s *session) ExecuteInternal(ctx context.Context, sql string) (recordSets []sqlexec.RecordSet, err error) {
+func (s *session) ExecuteInternal(ctx context.Context, sql string, args ...interface{}) (recordSets []sqlexec.RecordSet, err error) {
 	origin := s.sessionVars.InRestrictedSQL
 	s.sessionVars.InRestrictedSQL = true
 	defer func() {
 		s.sessionVars.InRestrictedSQL = origin
 	}()
-	return s.Execute(ctx, sql)
+
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("session.Execute", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+		ctx = opentracing.ContextWithSpan(ctx, span1)
+		logutil.Eventf(ctx, "execute: %s", sql)
+	}
+
+	stmtNodes, err := s.ParseWithParams(ctx, sql, args...)
+	if err != nil {
+		return nil, err
+	}
+	if len(stmtNodes) != 1 {
+		return nil, errors.New("ExecuteInternal() API doesn't support multiple statements")
+	}
+
+	rs, err := s.ExecuteStmt(ctx, stmtNodes[0])
+	if err != nil {
+		s.sessionVars.StmtCtx.AppendError(err)
+	}
+	if rs == nil {
+		return nil, err
+	}
+
+	return []sqlexec.RecordSet{rs}, err
```

Comment:
Since it does not support multiple statements, how about returning `sqlexec.RecordSet` to simplify the usage of this API?

Reply:
Maybe, but I would like to leave the refactor to other PRs. Maybe after fixing all injections.

Adding a varying arguments does not actually damage the compilation, but changing return value will.

Reply:
ok